Client/Frontend/Browser/TopTabsViewController.swift

```diff
@@ -174,42 +178,47 @@ class TopTabsViewController: UIViewController {
     }
     
     func newTabTapped() {
-        if let currentTab = tabManager.selectedTab, let index = tabsToDisplay.indexOf(currentTab),
-            let cell  = collectionView.cellForItemAtIndexPath(NSIndexPath(forItem: index, inSection: 0)) as? TopTabCell {
-            cell.selectedTab = false
-            if index > 0 {
-                cell.seperatorLine = true
-            }
+        dispatch_async(dispatch_get_main_queue()) {
+            self.delegate?.topTabsDidPressNewTab(self.isPrivate)
         }
-        delegate?.topTabsDidPressNewTab()
-        self.privateModeButton.enabled = false
-        collectionView.performBatchUpdates({ _ in
-            let count = self.collectionView.numberOfItemsInSection(0)
-            self.collectionView.insertItemsAtIndexPaths([NSIndexPath(forItem: count, inSection: 0)])
-            }, completion: { finished in
-                if finished {
-                    self.privateModeButton.enabled = true
-                    self.scrollToCurrentTab()
-                }
-        })
     }
-    
+
     func togglePrivateModeTapped() {
-        delegate?.topTabsDidPressPrivateModeButton(isPrivate ? lastNormalTab : lastPrivateTab)
+        if isUpdating || pendingReloadData {
+            return
+        }
+        delegate?.topTabsDidPressPrivateModeButton()
```

Comment:
The name of this delegate method isn't true if we're returning early. Naming it `topTabsDidTogglePrivateMode` would be better since the naming suggests that it gets called anytime the user taps the private mode button.